retention_rule.go

```diff
@@ -0,0 +1,315 @@
+	for _, r := range rules {
+		if !r.CutoffTime().IsZero() && r.CutoffTime().Before(from) {
+			// We've reached a rule that end before the start of the query - no more rules will apply
+			p.log.Debugf("rule cutoff time %s before query start %s, stopping build", r.CutoffTime(), from)
+			break
+		}
+
+		if !r.CutoverTime().IsZero() && r.CutoverTime().After(until) {
+			// We've reached a rule that doesn't apply by the time the query ends - keep going until we find
+			// an older rule which might apply
+			p.log.Debugf("rule cutover time %s after query end %s, skipping rule", r.CutoverTime(), until)
+			continue
+		}
+
+		// Bound the application of the rule by the query range
+		ruleStart, ruleEnd := from, until
+		if !r.CutoffTime().IsZero() && r.CutoffTime().Before(ruleEnd) {
+			ruleEnd = r.CutoffTime()
+		}
+
+		if !r.CutoverTime().IsZero() && r.CutoverTime().After(ruleStart) {
+			ruleStart = r.CutoverTime()
+		}
+
+		p.log.Debugf("rule applies from %s until %s", ruleStart, ruleEnd)
+
+		// Determine queries based on the retention policies in the rule.  Assumes
+		// that the retention policies are ordered by retention period, shortest
+		// retention period first.  Stops as soon as we've reached the last policy
+		// whose retention period falls within the rule or query bounds
+		lastPolicyStart := ruleEnd
+		policies := r.RetentionPolicies()
+		for n, policy := range policies {
+			query := query{
+				RetentionPolicy: policy,
+				Range: xtime.Range{
+					End:   lastPolicyStart,
+					Start: now.Add(-policy.RetentionPeriod().Duration()),
+				},
+			}
+
+			if !query.Range.Start.Before(ruleEnd) {
+				p.log.Debugf("policy %d (%s:%s) starts (%s) after end of rule (%s), does not apply",
+					n, policy.Resolution(), policy.RetentionPeriod(), query.Range.Start, ruleEnd)
+				continue
+			}
+
+			if query.Range.End.Before(ruleStart) {
+				p.log.Debugf("policy %d (%s:%s) ends (%s) before the start of the rule (%s), no more policies apply",
+					n, policy.Resolution(), policy.RetentionPeriod(), query.Range.End, ruleStart)
+				break
+			}
+
+			if query.Range.Start.Before(ruleStart) {
+				query.Range.Start = ruleStart
+			}
+
+			if query.Range.End.After(ruleEnd) {
+				query.Range.End = ruleEnd
+			}
+
+			p.log.Debugf("policy %d (%s:%s) applies from %s until %s (%s)",
+				n, policy.Resolution(), policy.RetentionPeriod(),
+				query.Range.Start, query.Range.End, query.Range.End.Sub(query.Range.Start))
+
+			if query.Range.IsEmpty() {
+				p.log.Debugf("skipping policy %d; does not apply within query range", n)
+				continue
+			}
+
+			lastPolicyStart = query.Range.Start
+			queries = append(queries, query)
+		}
+	}
+
+	if len(queries) == 0 {
+		return queries, nil
+	}
+
+	// Coalesce queries for the same retention policy
+	lastQuery := queries[0]
+	coalescedQueries := make([]query, 0, len(queries))
+	for i := 1; i < len(queries); i++ {
+		if queries[i].RetentionPolicy.Equal(lastQuery.RetentionPolicy) {
```

Comment:
This checks the properties of the retention policy but doesn't check the "ID" or pointer, are we worried about two retention policies that have the same attributes but aren't the same backing data source returning truthy here?